### PR TITLE
[IR] Support integer subbyte

### PR DESCRIPTION
**Pull request by xiaocenxiaocen**
- support sub byte integers in Hidet
```
a = register_tensor("int4b", [4, 4])
b = a[0, 2]
a[2, 2] = int4b(-5)
ptr = &a[0, 0]
ptr = ptr + 8
```

### CODE REVIEW

**Comment by yaoyaoding** [comment]
Hi @xiaocenxiaocen, let me know when the PR is ready to be reviewed, thanks!

**Comment by xiaocenxiaocen** [comment]
> Hi @xiaocenxiaocen, let me know when the PR is ready to be reviewed, thanks!

Sure. I will work on this in this week and the next week.

**Comment by xiaocenxiaocen** [comment]
Hi, @yaoyaoding. This PR is ready for review. Please take a look at it. Thanks.

**Comment by xiaocenxiaocen** [comment]
hidet-ci launch

File: python/hidet/transforms/lower_integer_subbyte.py [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+
+    def _set_subbyte_value(self, dtype: DataType, base: Var, offset: Expr, value: Expr):
+        storage_ty = dtype.storage
+        storage_bits = storage_ty.nbits
+        dtype_bits = dtype.nbits
+        divisor = storage_bits // dtype_bits
+        if divisor & (divisor - 1) != 0:
+            raise TypeError(f"data type not supported yet(got:{dtype})")
+        idx = simplify(offset // divisor)
+        offset_ = simplify(offset & (divisor - 1))
```
Comment (yaoyaoding) [inline python/hidet/transforms/lower_integer_subbyte.py:131]: I usually write `offset % divisor` directly and let the compiler to perform the optimization. But it's okay if you find this is more readable or comfortable to you.

File: tests/ir/test_int_subbyte.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+    func = script_module.build()
+    import torch
+
+    data = torch.empty((4, 4), dtype=torch.float32, device="cuda")
+    data = hidet.from_torch(data)
+    inp = torch.empty((4, 2), dtype=torch.int8, device="cuda")
+    inp = hidet.from_torch(inp)
+    func(data, inp)
+    print(data.cpu().numpy())
```
Comment (yaoyaoding) [inline tests/ir/test_int_subbyte.py:73]: Consider using something asserion here to check the correctness by comparing with some reference implementation (e.g., use pytorch to implement the same functionality).

File: tests/ir/test_int_subbyte.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+                        d = data[i, j]
+                        out[i, j] = f32(d)
+
+    func = script_module.build()
+    import torch
+
+    data = torch.empty((2, 2), dtype=torch.float32, device="cuda")
+    data = hidet.from_torch(data)
+    func(data)
+    print(data.cpu().numpy())
```
Comment (yaoyaoding) [inline tests/ir/test_int_subbyte.py:116]: Same here.

**Review by yaoyaoding — COMMENTED** [review]
Thanks @xiaocenxiaocen for the support of sub-integer type!

It looks good to me overall. And I left some minor suggestions to make some part more consistent with the existing implementation (like the data type). 

Feel free to merge this PR by yourself after you resolve those comments.

**Comment by hjjq** [comment]
$hidet-ci launch

File: python/hidet/ir/type.py [hunk 4]
```diff
@@ -67,12 +67,21 @@ def as_data_type(self) -> Optional[DataType]:
 class DataType(BaseType):
     """
     The data type that defines how to interpret the data in memory.
+
+    Note:
+    1. The _storage field for non-subbyte types is the type itself, while the _storage
+    for subbyte types is the type of its actual storage. e.g., the storage for int4b is uint8
+    2. The _storage field will be overwritten during the construction of subbyte types
+    2. The _nbits field in the constructor denotes the bit length of the storage, and
```
[the remark this answers is not in the document]
Comment (xiaocenxiaocen) [reply python/hidet/ir/type.py:75]: done

File: python/hidet/ir/type.py [hunk 5]
```diff
@@ -129,8 +138,21 @@ def short_name(self) -> str:
...
+            raise TypeError(f"Cannot access nbytes property for the type({self}")
         return self._nbytes
 
+    @property
+    def nbits(self) -> int:
+        return self._nbits
+
+    @property
+    def storage(self) -> DataType:
+        return self._storage
```
[the remark this answers is not in the document]
Comment (xiaocenxiaocen) [reply python/hidet/ir/type.py:151]: done

File: python/hidet/ir/type.py [hunk 6]
```diff
@@ -67,12 +67,21 @@ def as_data_type(self) -> Optional[DataType]:
...
+    for subbyte types is the type of its actual storage. e.g., the storage for int4b is uint8
+    2. The _storage field will be overwritten during the construction of subbyte types
+    2. The _nbits field in the constructor denotes the bit length of the storage, and
+    it will be overwritten in the constructor of subbyte types
     """
 
     def __init__(self, name: str, short_name: str, nbytes: int):
         self._name: str = name
         self._short_name: str = short_name
+        self._storage = self
```
[the remark this answers is not in the document]
Comment (xiaocenxiaocen) [reply python/hidet/ir/type.py:82]: done

File: python/hidet/ir/type.py [hunk 7]
```diff
@@ -129,8 +138,21 @@ def short_name(self) -> str:
...
+    @property
+    def nbits(self) -> int:
+        return self._nbits
+
+    @property
+    def storage(self) -> DataType:
+        return self._storage
+
+    def is_integer_subbyte(self) -> bool:
+        return self.is_integer() and self._nbits < 8
```
[the remark this answers is not in the document]
Comment (xiaocenxiaocen) [reply python/hidet/ir/type.py:154]: done

File: python/hidet/ir/type.py [hunk 8]
```diff
@@ -67,12 +67,21 @@ def as_data_type(self) -> Optional[DataType]:
 class DataType(BaseType):
     """
     The data type that defines how to interpret the data in memory.
+
+    Note:
+    1. The _storage field for non-subbyte types is the type itself, while the _storage
+    for subbyte types is the type of its actual storage. e.g., the storage for int4b is uint8
+    2. The _storage field will be overwritten during the construction of subbyte types
```
[the remark this answers is not in the document]
Comment (xiaocenxiaocen) [reply python/hidet/ir/type.py:74]: done

File: python/hidet/ir/type.py [hunk 9]
```diff
@@ -67,12 +67,21 @@ def as_data_type(self) -> Optional[DataType]:
...
+    2. The _nbits field in the constructor denotes the bit length of the storage, and
+    it will be overwritten in the constructor of subbyte types
     """
 
     def __init__(self, name: str, short_name: str, nbytes: int):
         self._name: str = name
         self._short_name: str = short_name
+        self._storage = self
         self._nbytes: int = nbytes
+        self._nbits: int = self._nbytes * 8
```
[the remark this answers is not in the document]
Comment (xiaocenxiaocen) [reply python/hidet/ir/type.py:84]: done

File: python/hidet/transforms/lower_integer_subbyte.py [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+        mask = storage_ty.constant(dtype.bits_mask)
+        return (base[idx] >> (offset_ * dtype_bits)) & mask
+
+    def _set_subbyte_value(self, dtype: DataType, base: Var, offset: Expr, value: Expr):
+        storage_ty = dtype.storage
+        storage_bits = storage_ty.nbits
+        dtype_bits = dtype.nbits
+        divisor = storage_bits // dtype_bits
+        if divisor & (divisor - 1) != 0:
```
[the remark this answers is not in the document]
Comment (xiaocenxiaocen) [reply python/hidet/transforms/lower_integer_subbyte.py:128]: done

File: python/hidet/transforms/lower_integer_subbyte.py (hunk 1 above)
Comment (xiaocenxiaocen) [reply python/hidet/transforms/lower_integer_subbyte.py:131]: done

**Comment by xiaocenxiaocen** [comment]
$hidet-ci launch

**Comment by xiaocenxiaocen** [comment]
$hidet-ci launch

**Pull request merged**